In [1]:
import pandas as pd 
import locale
import calendar
import utils as ut

Aca me ubico en la carpeta de los archivos necesarios

In [3]:
%cd C:\\Users\\Familia Lara Pernia\\Desktop\\Proyecto2_SiniestrosViales_Henry
%pwd

C:\Users\Familia Lara Pernia\Desktop\Proyecto2_SiniestrosViales_Henry


'C:\\Users\\Familia Lara Pernia\\Desktop\\Proyecto2_SiniestrosViales_Henry'

Cargo archivos a trabajar

In [5]:
hechos = pd.read_excel("data/excel/homicidios.xlsx", sheet_name="HECHOS")
hechos_dict = pd.read_excel("data/excel/homicidios.xlsx", sheet_name="DICCIONARIO_HECHOS")

Tratamiento de los datos del dataframe 'hechos' y 'hechos_dict'

In [6]:
hechos

ID  N_VICTIMAS      FECHA  AAAA  MM  DD      HORA  HH  \
0    2016-0001           1 2016-01-01  2016   1   1  04:00:00   4   
1    2016-0002           1 2016-01-02  2016   1   2  01:15:00   1   
2    2016-0003           1 2016-01-03  2016   1   3  07:00:00   7   
3    2016-0004           1 2016-01-10  2016   1  10  00:00:00   0   
4    2016-0005           1 2016-01-21  2016   1  21  05:20:00   5   
..         ...         ...        ...   ...  ..  ..       ...  ..   
691  2021-0093           1 2021-12-13  2021  12  13  17:10:00  17   
692  2021-0094           1 2021-12-20  2021  12  20  01:10:00   1   
693  2021-0095           1 2021-12-30  2021  12  30  00:43:00   0   
694  2021-0096           1 2021-12-15  2021  12  15  10:30:00  10   
695  2021-0097           1 2021-11-18  2021  11  18  06:10:00   6   

                               LUGAR_DEL_HECHO TIPO_DE_CALLE  ...  Altura  \
0    AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  ...     NaN   
1            AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ  ...     NaN   
2                           AV ENTRE RIOS 2034       AVENIDA  ...  2034.0   
3        AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA  ...     NaN   
4    AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA  ...     NaN   
..                                         ...           ...  ...     ...   
691                          AV. RIESTRA Y MOM       AVENIDA  ...     NaN   
692                    AU DELLEPIANE Y LACARRA     AUTOPISTA  ...     NaN   
693                        AV. GAONA Y TERRADA       AVENIDA  ...     NaN   
694                         AV. EVA PERON 4071       AVENIDA  ...  4071.0   
695                    PADRE CARLOS MUJICA 709         CALLE  ...   709.0   

                                   Cruce  \
0    FERNANDEZ DE LA CRUZ, F., GRAL. AV.   
1                    DE LOS CORRALES AV.   
2                                    NaN   
3               VILLEGAS, CONRADO, GRAL.   
4                SAENZ PE?A, LUIS, PRES.   
..                                   ...   
691                                  MOM   
692                          LACARRA AV.   
693                              TERRADA   
694                                  NaN   
695                                  NaN   

                                 Dirección Normalizada COMUNA  \
0    PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...      8   
1                 PAZ, GRAL. AV. y DE LOS CORRALES AV.      9   
2                                  ENTRE RIOS AV. 2034      1   
3            LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.      8   
4              SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.      1   
..                                                 ...    ...   
691                                  RIESTRA AV. y MOM      7   
692         DELLEPIANE, LUIS, TTE. GRAL. y LACARRA AV.      9   
693                                GAONA AV. y TERRADA     11   
694                                PERON, EVA AV. 4071      9   
695                            PADRE CARLOS MUGICA 709      1   

                                   XY (CABA)         pos x         pos y  \
0      Point (98896.78238426 93532.43437792)  -58.47533969  -34.68757022   
1      Point (95832.05571093 95505.41641999)  -58.50877521  -34.66977709   
2     Point (106684.29090040 99706.57687843)  -58.39040293  -34.63189362   
3      Point (99840.65224780 94269.16534422)  -58.46503904  -34.68092974   
4    Point (106980.32827929 100752.16915795)  -58.38718297  -34.62246630   
..                                       ...           ...           ...   
691   Point (102728.60090138 98186.24929177)  -58.43353773  -34.64561636   
692    Point (99624.29795829 97569.69801131)  -58.46739825  -34.65117757   
693   Point (99116.45492358 101045.23284826)  -58.47293407  -34.61984745   
694    Point (99324.54463985 97676.26932409)  -58.47066794  -34.65021673   
695  Point (107664.16647795 104708.63962087)  -58.37976155  -34.58679619   

      PARTICIPANTES    VICTIMA    A

In [8]:
hechos_dict

variables y definiciones  \
0                        ID   
1                N_VICTIMAS   
2                     FECHA   
3                      AAAA   
4                        MM   
5                        DD   
6                      HORA   
7                        HH   
8           LUGAR_DEL_HECHO   
9             TIPO_DE_CALLE   
10                    Calle   
11                   Altura   
12                    Cruce   
13    Dirección Normalizada   
14                   COMUNA   
15                XY (CABA)   
16                    pos x   
17                    pos y   
18            PARTICIPANTES   
19                  VICTIMA   
20                  ACUSADO   
21                      NaN   
22                Variables   
23            TIPO_DE_CALLE   
24                      NaN   
25                      NaN   
26                      NaN   
27                  VICTIMA   
28                      NaN   
29                      NaN   
30                      NaN   
31                      NaN   
32                      NaN   
33                      NaN   
34                      NaN   
35                      NaN   
36                  ACUSADO   
37                      NaN   
38                      NaN   
39                      NaN   
40                      NaN   
41                      NaN   
42                      NaN   
43                      NaN   
44                      NaN   
45            PARTICIPANTES   

                                          descripción  \
0                   identificador unico del siniestro   
1                                cantidad de víctimas   
2                         fecha en formato dd/mm/aaaa   
3                                                 año   
4                                                 mes   
5                                         día del mes   
6                                  hora del siniestro   
7                               franja horaria entera   
8                                 Dirección del hecho   
9   Tipo de arteria. En el caso de intersecciones ...   
10     nombre de la arteria donde se produjo el hecho   
11     altura de la arteria donde se produjo el hecho   
12           cruce en caso de que sea una encrucijada   
13              direccion en formato normalizado USIG   
14                       Comuna de la ciudad (1 a 15)   
15                              geocodificación plana   
16                longitud con separador punto. WGS84   
17                 latitud con separador punto. WGS84   
18                    conjunción de víctima y acusado   
19  Vehículo que ocupaba quien haya fallecido a se...   
20  Vehículo que ocupaba quien resultó acusado/a d...   
21                                                NaN   
22                                            valores   
23                                              calle   
24                                            avenida   
25                                          autopista   
26                                        general paz   
27                                             PEATON   
28                                               MOTO   
29                                               AUTO   
30                                             CARGAS   
31                                          BICICLETA   
32                                          PASAJEROS   
33                                              MOVIL   
34                                               OTRO   
35                                                 SD   
36                                               AUTO   
37                                          BICICLETA   
38                                             CARGAS   
39                                               MOTO   
40                                        OBJETO FIJO   
41                                          PASAJEROS   
42                                               TREN   
43                                   

Ahora, necesitamos ver un resumen sobre la base de datos de hechos

In [9]:
ut.data_summ(hechos)


Total filas:  696

Total filas nulas:  0

Total filas duplicadas:  0


Columnas                                          Tipo_Dato  \
0                      ID                                    [<class 'str'>]   
1              N_VICTIMAS                                    [<class 'int'>]   
2                   FECHA  [<class 'pandas._libs.tslibs.timestamps.Timest...   
3                    AAAA                                    [<class 'int'>]   
4                      MM                                    [<class 'int'>]   
5                      DD                                    [<class 'int'>]   
6                    HORA  [<class 'datetime.time'>, <class 'str'>, <clas...   
7                      HH                     [<class 'int'>, <class 'str'>]   
8         LUGAR_DEL_HECHO                                    [<class 'str'>]   
9           TIPO_DE_CALLE                                    [<class 'str'>]   
10                  Calle                   [<class 'str'>, <class 'float'>]   
11                 Altura                                  [<class 'float'>]   
12                  Cruce                   [<class 'str'>, <class 'float'>]   
13  Dirección Normalizada                   [<class 'str'>, <class 'float'>]   
14                 COMUNA                                    [<class 'int'>]   
15              XY (CABA)                                    [<class 'str'>]   
16                  pos x                                    [<class 'str'>]   
17                  pos y                                    [<class 'str'>]   
18          PARTICIPANTES                                    [<class 'str'>]   
19                VICTIMA                                    [<class 'str'>]   
20                ACUSADO                                    [<class 'str'>]   

    No_miss_Qty  %Missing  Missing_Qty  
0           696      0.00            0  
1           696      0.00            0  
2           696      0.00            0  
3           696      0.00            0  
4           696      0.00            0  
5           696      0.00            0  
6           696      0.00            0  
7           696      0.00            0  
8           696      0.00            0  
9           696      0.00            0  
10          695      0.14            1  
11          129     81.47          567  
12          525     24.57          171  
13          688      1.15            8  
14          696      0.00            0  
15          696      0.00            0  
16          696      0.00            0  
17          696      0.00            0  
18          696      0.00            0  
19          696      0.00            0  
20          696      0.00            0

Nuestro resumen de datos muestra algunas columnas en mayúsculas y otras en minúsculas. También muestra algunas columnas con datos nulos y algunas columnas con diferentes tipos de datos. Nuestro primer paso será estandarizar la escritura de columnas, analizar y transformar datos nulos (si es posible) y luego analizar columnas con diferentes tipos de datos.

Estandarizar formato de columnas

In [10]:
# Reemplazar guiones bajos con espacios
hechos.columns = hechos.columns.str.replace('_', ' ')
# La primera letra está en mayúscula
hechos.columns = hechos.columns.str.title()
# Cambiar el nombre de las columnas
hechos = hechos.rename(columns={'Id': 'Hechos_Id',                        
                                        'N Victimas': 'Número De Victimas',
                                        'Fecha': 'Fecha',
                                        'Aaaa':'Año',
                                        'Mm':'Mes',
                                        'Dd':'Dia',
                                        'Hh':'Franja Horaria',
                                        'Xy (Caba)':'XY (CABA)',
                                        'Victima': 'Víctima'})


hechos.columns

Index(['Hechos_Id', 'Número De Victimas', 'Fecha', 'Año', 'Mes', 'Dia', 'Hora',
       'Franja Horaria', 'Lugar Del Hecho', 'Tipo De Calle', 'Calle', 'Altura',
       'Cruce', 'Dirección Normalizada', 'Comuna', 'XY (CABA)', 'Pos X',
       'Pos Y', 'Participantes', 'Víctima', 'Acusado'],
      dtype='object')

Columnas con datos nulos

* La columna 'Calle' tiene 1 dato nulo
* La columna 'Altura' tiene 567 datos nulos
* La columna 'Cruce' tiene 171 datos nulos
* La columna 'Dirección Normalizada' tiene 8 datos nulos

Podemos ver en la tabla que la columna 'Lugar del hecho' está relacionada con las columnas 'Calle', 'Cruce' y 'Dirección Normalizada'. Pasamos a analizar este conjunto de columnas para decidir qué hacer con los datos nulos.

In [11]:
hechos[hechos['Calle'].isnull()][['Dirección Normalizada', 'Cruce','Altura', 'Calle', 'Lugar Del Hecho']]

Dirección Normalizada Cruce  Altura Calle Lugar Del Hecho
119                   NaN   NaN     NaN   NaN              SD

In [12]:
hechos[hechos['Calle'].isnull()]

Hechos_Id  Número De Victimas      Fecha   Año  Mes  Dia      Hora  \
119  2016-0151                   1 2016-11-18  2016   11   18  20:35:00   

    Franja Horaria Lugar Del Hecho Tipo De Calle  ... Altura  Cruce  \
119             20              SD         CALLE  ...    NaN    NaN   

    Dirección Normalizada Comuna    XY (CABA) Pos X Pos Y Participantes  \
119                   NaN      0  Point (. .)     .     .     PEATON-SD   

    Víctima Acusado  
119  PEATON      SD  

[1 rows x 21 columns]

Vemos que la columna 'Calle' tiene solo un valor nulo. Al analizar toda la fila, nos damos cuenta que a esta fila le falta información relevante, por lo tanto decidimos eliminarla.

In [13]:
hechos = hechos.dropna(subset=['Calle']).reset_index(drop=True)

Ahora analizamos la columna 'Altura'

In [14]:
hechos[hechos['Altura'].notna()][['Dirección Normalizada', 'Cruce','Altura', 'Calle', 'Lugar Del Hecho']]

Dirección Normalizada Cruce  Altura                Calle  \
2        ENTRE RIOS AV. 2034   NaN  2034.0       ENTRE RIOS AV.   
9        ENTRE RIOS AV. 1366   NaN  1366.0       ENTRE RIOS AV.   
14              SUIPACHA 156   NaN   156.0             SUIPACHA   
33                 LIMA 1483   NaN  1483.0                 LIMA   
42    CAMPOS, LUIS M. AV. 30   NaN    30.0  CAMPOS, LUIS M. AV.   
..                       ...   ...     ...                  ...   
676             TUCUMAN 3112   NaN  3112.0              TUCUMAN   
684           ESMERALDA 1359   NaN  1359.0            ESMERALDA   
686           GAONA AV. 3655   NaN  3655.0            GAONA AV.   
693      PERON, EVA AV. 4071   NaN  4071.0       PERON, EVA AV.   
694  PADRE CARLOS MUGICA 709   NaN   709.0  PADRE CARLOS MUJICA   

             Lugar Del Hecho  
2         AV ENTRE RIOS 2034  
9         AV ENTRE RIOS 1366  
14              SUIPACHA 156  
33                 LIMA 1483  
42   AV LUIS MARIA CAMPOS 30  
..                       ...  
676             TUCUMAN 3112  
684           ESMERALDA 1359  
686           AV. GAONA 3655  
693       AV. EVA PERON 4071  
694  PADRE CARLOS MUJICA 709  

[129 rows x 5 columns]

Analizando los datos de la columna 'Altura' se procede a eliminarla debido a la gran cantidad de ítems faltantes (81,47%) y no habiendo forma de completar las columnas; Investigando los datos de la columna 'Altura' hemos visto que se refiere a la numeración de la calle donde se produce el accidente. El conjunto de datos nos muestra que la mayoría de los accidentes ocurren en intersecciones o esquinas. De hecho, esta información se encuentra en la columna 'Dirección Normalizada'

In [15]:
# Borramos columna 'Altura'
hechos= hechos.drop('Altura', axis=1)

Ahora analizamos la columna 'Cruce'

In [16]:
hechos[hechos['Cruce'].isnull()][['Dirección Normalizada', 'Cruce', 'Calle', 'Lugar Del Hecho']]

Dirección Normalizada Cruce  \
2                           ENTRE RIOS AV. 2034   NaN   
9                           ENTRE RIOS AV. 1366   NaN   
14                                 SUIPACHA 156   NaN   
33                                    LIMA 1483   NaN   
35   AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI   NaN   
..                                          ...   ...   
679                AUTOPISTA 25 DE MAYO y PASCO   NaN   
684                              ESMERALDA 1359   NaN   
686                              GAONA AV. 3655   NaN   
693                         PERON, EVA AV. 4071   NaN   
694                     PADRE CARLOS MUGICA 709   NaN   

                                          Calle  \
2                                ENTRE RIOS AV.   
9                                ENTRE RIOS AV.   
14                                     SUIPACHA   
33                                         LIMA   
35   AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI   
..                                          ...   
679                        AUTOPISTA 25 DE MAYO   
684                                   ESMERALDA   
686                                   GAONA AV.   
693                              PERON, EVA AV.   
694                         PADRE CARLOS MUJICA   

                                       Lugar Del Hecho  
2                                   AV ENTRE RIOS 2034  
9                                   AV ENTRE RIOS 1366  
14                                        SUIPACHA 156  
33                                           LIMA 1483  
35   AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI KM....  
..                                                 ...  
679                   AU 25 DE MAYO KM7 (ALTURA PASCO)  
684                                     ESMERALDA 1359  
686                                     AV. GAONA 3655  
693                                 AV. EVA PERON 4071  
694                            PADRE CARLOS MUJICA 709  

[170 rows x 4 columns]

In [17]:
hechos[hechos['Cruce'].notna()][['Dirección Normalizada', 'Cruce', 'Calle', 'Lugar Del Hecho']]

Dirección Normalizada  \
0    PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...   
1                 PAZ, GRAL. AV. y DE LOS CORRALES AV.   
3            LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.   
4              SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.   
5                     27 DE FEBRERO AV. y ESCALADA AV.   
..                                                 ...   
688                           BAIGORRIA y HUGO, VICTOR   
689                     RIVADAVIA AV. y PUEYRREDON AV.   
690                                  RIESTRA AV. y MOM   
691         DELLEPIANE, LUIS, TTE. GRAL. y LACARRA AV.   
692                                GAONA AV. y TERRADA   

                                   Cruce                         Calle  \
0    FERNANDEZ DE LA CRUZ, F., GRAL. AV.              PIEDRA BUENA AV.   
1                    DE LOS CORRALES AV.                PAZ, GRAL. AV.   
3               VILLEGAS, CONRADO, GRAL.                LARRAZABAL AV.   
4                SAENZ PE?A, LUIS, PRES.                  SAN JUAN AV.   
5                           ESCALADA AV.             27 DE FEBRERO AV.   
..                                   ...                           ...   
688                         HUGO, VICTOR                     BAIGORRIA   
689                       PUEYRREDON AV.                 RIVADAVIA AV.   
690                                  MOM                   RIESTRA AV.   
691                          LACARRA AV.  DELLEPIANE, LUIS, TTE. GRAL.   
692                              TERRADA                     GAONA AV.   

                               Lugar Del Hecho  
0    AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ  
1            AV GRAL PAZ  Y AV DE LOS CORRALES  
3        AV LARRAZABAL Y GRAL VILLEGAS CONRADO  
4    AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA   
5               AV 27 DE FEBRERO Y AV ESCALADA  
..                                         ...  
688                    BAIGORRIA Y VICTOR HUGO  
689             AV. RIVADAVIA Y AV. PUEYRREDON  
690                          AV. RIESTRA Y MOM  
691                    AU DELLEPIANE Y LACARRA  
692                        AV. GAONA Y TERRADA  

[525 rows x 4 columns]

Observamos que la columna 'Cruce' hace referencia a las intersecciones entre dos calles y tiene datos sólo cuando el accidente está referenciado en la intersección de dos calles. De lo contrario es nulo.

Cuando el accidente es en un cruce de calles, los datos de la columna 'Cruce' se encuentran en la columna 'LUGAR_DEL_HECHO', por lo tanto, vamos a definir valores booleanos para esta columna para cuando el accidente se establezca en un cruce de calles. 

In [18]:
hechos['Cruce'] = hechos['Cruce'].notna()

In [19]:
hechos[hechos['Dirección Normalizada'].isnull()][['Dirección Normalizada', 'Cruce', 'Calle', 'Lugar Del Hecho']]

Dirección Normalizada  Cruce                                 Calle  \
38                    NaN  False                 LUGONES, LEOPOLDO AV.   
106                   NaN  False     AUTOPISTA BUENOS AIRES - LA PLATA   
179                   NaN  False               AUTOPISTA PERITO MORENO   
180                   NaN  False  AUTOPISTA DELLEPIANE LUIS TTE. GRAL.   
312                   NaN  False                 LUGONES, LEOPOLDO AV.   
545                   NaN  False                 LUGONES, LEOPOLDO AV.   
620                   NaN  False     AUTOPISTA BUENOS AIRES - LA PLATA   

                             Lugar Del Hecho  
38                AUTOPISTA LUGONES PK 10000  
106         AU BUENOS AIRES - LA PLATA KM. 4  
179  AU PERITO MORENO Y RAMAL ENLACE AU1/AU6  
180                       AU DELLEPIANE 2400  
312                 AUTOPISTA LUGONES KM 4.7  
545             LUGONES, LEOPOLDO AV. KM 6,1  
620         AU BUENOS AIRES LA PLATA  KM 4,5

In [20]:
hechos['Dirección Normalizada'].fillna('SD', inplace=True)

In [21]:
hechos.head()

Hechos_Id  Número De Victimas      Fecha   Año  Mes  Dia      Hora  \
0  2016-0001                   1 2016-01-01  2016    1    1  04:00:00   
1  2016-0002                   1 2016-01-02  2016    1    2  01:15:00   
2  2016-0003                   1 2016-01-03  2016    1    3  07:00:00   
3  2016-0004                   1 2016-01-10  2016    1   10  00:00:00   
4  2016-0005                   1 2016-01-21  2016    1   21  05:20:00   

  Franja Horaria                            Lugar Del Hecho Tipo De Calle  \
0              4  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA   
1              1          AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ   
2              7                         AV ENTRE RIOS 2034       AVENIDA   
3              0      AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA   
4              5  AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA   

              Calle  Cruce                              Dirección Normalizada  \
0  PIEDRA BUENA AV.   True  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...   
1    PAZ, GRAL. AV.   True               PAZ, GRAL. AV. y DE LOS CORRALES AV.   
2    ENTRE RIOS AV.  False                                ENTRE RIOS AV. 2034   
3    LARRAZABAL AV.   True          LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.   
4      SAN JUAN AV.   True            SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.   

   Comuna                                XY (CABA)         Pos X  \
0       8    Point (98896.78238426 93532.43437792)  -58.47533969   
1       9    Point (95832.05571093 95505.41641999)  -58.50877521   
2       1   Point (106684.29090040 99706.57687843)  -58.39040293   
3       8    Point (99840.65224780 94269.16534422)  -58.46503904   
4       1  Point (106980.32827929 100752.16915795)  -58.38718297   

          Pos Y   Participantes Víctima    Acusado  
0  -34.68757022       MOTO-AUTO    MOTO       AUTO  
1  -34.66977709  AUTO-PASAJEROS    AUTO  PASAJEROS  
2  -34.63189362       MOTO-AUTO    MOTO       AUTO  
3  -34.68092974         MOTO-SD    MOTO         SD  
4  -34.62246630  MOTO-PASAJEROS    MOTO  PASAJEROS

Columnas con diferentes tipos de datos.


In [22]:
ut.data_summ(hechos)


Total filas:  695

Total filas nulas:  0

Total filas duplicadas:  0


Columnas                                          Tipo_Dato  \
0               Hechos_Id                                    [<class 'str'>]   
1      Número De Victimas                                    [<class 'int'>]   
2                   Fecha  [<class 'pandas._libs.tslibs.timestamps.Timest...   
3                     Año                                    [<class 'int'>]   
4                     Mes                                    [<class 'int'>]   
5                     Dia                                    [<class 'int'>]   
6                    Hora  [<class 'datetime.time'>, <class 'str'>, <clas...   
7          Franja Horaria                     [<class 'int'>, <class 'str'>]   
8         Lugar Del Hecho                                    [<class 'str'>]   
9           Tipo De Calle                                    [<class 'str'>]   
10                  Calle                                    [<class 'str'>]   
11                  Cruce                                   [<class 'bool'>]   
12  Dirección Normalizada                                    [<class 'str'>]   
13                 Comuna                                    [<class 'int'>]   
14              XY (CABA)                                    [<class 'str'>]   
15                  Pos X                                    [<class 'str'>]   
16                  Pos Y                                    [<class 'str'>]   
17          Participantes                                    [<class 'str'>]   
18                Víctima                                    [<class 'str'>]   
19                Acusado                                    [<class 'str'>]   

    No_miss_Qty  %Missing  Missing_Qty  
0           695       0.0            0  
1           695       0.0            0  
2           695       0.0            0  
3           695       0.0            0  
4           695       0.0            0  
5           695       0.0            0  
6           695       0.0            0  
7           695       0.0            0  
8           695       0.0            0  
9           695       0.0            0  
10          695       0.0            0  
11          695       0.0            0  
12          695       0.0            0  
13          695       0.0            0  
14          695       0.0            0  
15          695       0.0            0  
16          695       0.0            0  
17          695       0.0            0  
18          695       0.0            0  
19          695       0.0            0

Las columnas 'Hora' y 'Franja Horaria' tienen tipos de datos mixtos. Vamos a analizar y transformar (si es posible)

In [23]:
# Cantidad de valores por tipo de dato en la columna 'hora'
hechos['Hora'].apply(type).value_counts()

Hora
<class 'datetime.time'>        607
<class 'str'>                   85
<class 'datetime.datetime'>      3
Name: count, dtype: int64

Vamos a analizar columnas con tipo de datos 'str

In [24]:
datos_str = hechos.loc[hechos['Hora'].apply(lambda x: isinstance(x, str))]
datos_str

Hechos_Id  Número De Victimas      Fecha   Año  Mes  Dia      Hora  \
274  2018-0001                   1 2018-01-01  2018    1    1  22:50:00   
275  2018-0002                   1 2018-01-03  2018    1    3  16:40:00   
276  2018-0003                   1 2018-01-05  2018    1    5  10:12:00   
277  2018-0004                   1 2018-01-09  2018    1    9  09:40:00   
278  2018-0005                   1 2018-01-09  2018    1    9  22:30:00   
..         ...                 ...        ...   ...  ...  ...       ...   
411  2018-0138                   1 2018-12-21  2018   12   21  09:30:00   
412  2018-0139                   1 2018-12-22  2018   12   22  06:30:00   
415  2018-0142                   1 2018-12-25  2018   12   25  13:30:00   
416  2018-0143                   1 2018-12-29  2018   12   29  05:25:00   
517  2019-0103                   1 2019-12-18  2019   12   18        SD   

    Franja Horaria                           Lugar Del Hecho Tipo De Calle  \
274             22                     San Juan Av. Y Rincón       AVENIDA   
275             16                                  Peru 937         CALLE   
276             10    Pueyrredon, Honorio, Dr. Av. y Galicia       AVENIDA   
277              9           Del Libertador Av. y Manzanares       AVENIDA   
278             22                      Paseo Colon Av. 1019       AVENIDA   
..             ...                                       ...           ...   
411              9                             Bermudez 1893         CALLE   
412              6                            Santander 1974         CALLE   
415             13                       PAZ, GRAL. AV. 5172      GRAL PAZ   
416              5  Olivera Av. y Alberdi, Juan Bautista Av.       AVENIDA   
517             SD                   PAZ, GRAL. AV. Y GRIVEO      GRAL PAZ   

                            Calle  Cruce  \
274                  SAN JUAN AV.   True   
275                          PERU  False   
276  PUEYRREDON, HONORIO, DR. AV.   True   
277            DEL LIBERTADOR AV.   True   
278               PASEO COLON AV.  False   
..                            ...    ...   
411                      BERMUDEZ  False   
412                     SANTANDER  False   
415                PAZ, GRAL. AV.  False   
416                   OLIVERA AV.   True   
517                PAZ, GRAL. AV.   True   

                        Dirección Normalizada  Comuna  \
274                     SAN JUAN AV. y RINCON       3   
275                                  PERU 937       1   
276    PUEYRREDON, HONORIO, DR. AV. y GALICIA      15   
277           DEL LIBERTADOR AV. y MANZANARES      13   
278                      PASEO COLON AV. 1019       1   
..                                        ...     ...   
411                             BERMUDEZ 1893      10   
412                            SANTANDER 1974       7   
415                       PAZ, GRAL. AV. 5172      12   
416  OLIVERA AV. y ALBERDI, JUAN BAUTISTA AV.      10   
517                   PAZ, GRAL. AV. y GRIVEO      11   

                                   XY (CABA)         Pos X         Pos Y  \
274  Point (106229.70400707 100696.43939679)  -58.39536773  -34.62297346   
275  Point (108138.18410245 101140.78821350)  -58.37456088  -34.61895467   
276  Point (101348.13972450 102808.02053976)  -58.44860217  -34.60395657   
277  Point (100183.96103971 109533.84269558)  -58.46129513  -34.54332766   
278  Point (108641.50518777 101075.93727953)  -58.36907201  -34.61953517   
..                                       ...           ...           ...   
411   Point (95944.83421152 100945.68379559)  -58.50751833  -34.62073702   
412   Point (101510.09611838 98933.71958961)  -58.44682986  -34.63888084   
415   Point (95716.93710226 106295.47905158)  -58.50997604  -34.57251072   
416    Point (97912.79273192 98756.38729095)  -58.48606475  -34.64047818   
517   Point (94643.11254058 103831.57115061)  -58.52169422  -34.59471640   

           Participantes    Víctima    A

Vemos que los datos de tipo str tienen formato de hora, excepto uno que es SD. Convertiremos los datos a tipo de tiempo y realizaremos una imputación con los datos SD con referencia al modo

In [25]:
hechos['Hora'] = pd.to_datetime(hechos['Hora'], format='%H:%M:%S', errors='coerce').dt.time

In [26]:
hechos['Hora'].apply(type).value_counts()

Hora
<class 'datetime.time'>                          694
<class 'pandas._libs.tslibs.nattype.NaTType'>      1
Name: count, dtype: int64

In [27]:
# calcular la moda
hora_moda = hechos['Hora'].mode().iloc[0]
#  El valor None se reemplaza por la hora más común.
hechos['Hora'].fillna(hora_moda, inplace=True)
hechos.loc[517]

Hechos_Id                                             2019-0103
Número De Victimas                                            1
Fecha                                       2019-12-18 00:00:00
Año                                                        2019
Mes                                                          12
Dia                                                          18
Hora                                                   09:00:00
Franja Horaria                                               SD
Lugar Del Hecho                         PAZ, GRAL. AV. Y GRIVEO
Tipo De Calle                                          GRAL PAZ
Calle                                            PAZ, GRAL. AV.
Cruce                                                      True
Dirección Normalizada                   PAZ, GRAL. AV. y GRIVEO
Comuna                                                       11
XY (CABA)                Point (94643.11254058 103831.57115061)
Pos X                                   

In [29]:
datos_str = hechos.loc[hechos['Franja Horaria'].apply(lambda x: isinstance(x, str))]
datos_str

Hechos_Id  Número De Victimas      Fecha   Año  Mes  Dia      Hora  \
517  2019-0103                   1 2019-12-18  2019   12   18  09:00:00   

    Franja Horaria          Lugar Del Hecho Tipo De Calle           Calle  \
517             SD  PAZ, GRAL. AV. Y GRIVEO      GRAL PAZ  PAZ, GRAL. AV.   

     Cruce    Dirección Normalizada  Comuna  \
517   True  PAZ, GRAL. AV. y GRIVEO      11   

                                  XY (CABA)         Pos X         Pos Y  \
517  Point (94643.11254058 103831.57115061)  -58.52169422  -34.59471640   

    Participantes Víctima Acusado  
517     MOTO-MOTO    MOTO    MOTO

In [30]:
# Camnio de str a int
hechos['Franja Horaria'] = hechos['Franja Horaria'].apply(lambda x: int(hora_moda.hour) if x == "SD" else x)
# Verifico
hechos['Franja Horaria'].apply(type).value_counts()

Franja Horaria
<class 'int'>    695
Name: count, dtype: int64

In [31]:
ut.data_summ(hechos)


Total filas:  695

Total filas nulas:  0

Total filas duplicadas:  0


Columnas                                          Tipo_Dato  \
0               Hechos_Id                                    [<class 'str'>]   
1      Número De Victimas                                    [<class 'int'>]   
2                   Fecha  [<class 'pandas._libs.tslibs.timestamps.Timest...   
3                     Año                                    [<class 'int'>]   
4                     Mes                                    [<class 'int'>]   
5                     Dia                                    [<class 'int'>]   
6                    Hora                          [<class 'datetime.time'>]   
7          Franja Horaria                                    [<class 'int'>]   
8         Lugar Del Hecho                                    [<class 'str'>]   
9           Tipo De Calle                                    [<class 'str'>]   
10                  Calle                                    [<class 'str'>]   
11                  Cruce                                   [<class 'bool'>]   
12  Dirección Normalizada                                    [<class 'str'>]   
13                 Comuna                                    [<class 'int'>]   
14              XY (CABA)                                    [<class 'str'>]   
15                  Pos X                                    [<class 'str'>]   
16                  Pos Y                                    [<class 'str'>]   
17          Participantes                                    [<class 'str'>]   
18                Víctima                                    [<class 'str'>]   
19                Acusado                                    [<class 'str'>]   

    No_miss_Qty  %Missing  Missing_Qty  
0           695       0.0            0  
1           695       0.0            0  
2           695       0.0            0  
3           695       0.0            0  
4           695       0.0            0  
5           695       0.0            0  
6           695       0.0            0  
7           695       0.0            0  
8           695       0.0            0  
9           695       0.0            0  
10          695       0.0            0  
11          695       0.0            0  
12          695       0.0            0  
13          695       0.0            0  
14          695       0.0            0  
15          695       0.0            0  
16          695       0.0            0  
17          695       0.0            0  
18          695       0.0            0  
19          695       0.0            0

Datos Geográficos

Ahora analicemos las columnas correspondientes a la ubicación geográfica de los accidentes:

In [32]:
unique_values_X = hechos['Pos X'].unique()

# Mostrar valores unicos
print(unique_values_X)

['-58.47533969' '-58.50877521' '-58.39040293' '-58.46503904'
 '-58.38718297' '-58.44451316' '-58.50095869' '-58.38048577'
 '-58.39114932' '-58.43760020' '-58.36882665' '-58.44989671'
 '-58.50990884' '-58.37952223' '-58.45157703' '-58.38144696'
 '-58.46244952' '-58.49216191' '-58.44497874' '-58.38199706'
 '-58.47378664' '-58.47064556' '-58.40623949' '-58.39228965'
 '-58.37813396' '-58.46275892' '-58.49683458' '-58.40603217'
 '-58.46076550' '-58.47832463' '-58.43213235' '-58.46743471'
 '-58.38247061' '-58.44759966' '-58.37714647568196' '-58.47305463' '.'
 '-58.48429180' '-58.48695354' '-58.40021959' '-58.42957885'
 '-58.38247542' '-58.38222219' '-58.49485242' '-58.45227946'
 '-58.48455411' '-58.39056593' '-58.39118948' '-58.49021952'
 '-58.49896559' '-58.47461076' '-58.46210891' '-58.37027770'
 '-58.40491455' '-58.45504154' '-58.52927982' '-58.48721060'
 '-58.42273092' '-58.44788671' '-58.38110563' '-58.42378834'
 '-58.48727942' '-58.37864583' '-58.50151750' '-58.43233425'
 '-58.41971993

In [33]:
hechos[hechos['Pos X'] == '.']

Hechos_Id  Número De Victimas      Fecha   Año  Mes  Dia      Hora  \
38   2016-0052                   1 2016-04-20  2016    4   20  20:00:00   
106  2016-0136                   1 2016-10-25  2016   10   25  00:00:00   
138  2016-0174                   1 2016-12-27  2016   12   27  00:00:00   
175  2017-0042                   1 2017-04-10  2017    4   10  09:00:00   
179  2017-0050                   2 2017-04-28  2017    4   28  11:08:08   
180  2017-0051                   1 2017-05-01  2017    5    1  03:47:47   
255  2017-0140                   1 2017-11-19  2017   11   19  23:22:17   
312  2018-0039                   1 2018-04-21  2018    4   21  22:15:00   
545  2020-0026                   1 2020-05-17  2020    5   17  06:40:00   
558  2020-0039                   1 2020-09-01  2020    9    1  19:17:42   
620  2021-0023                   1 2021-03-01  2021    3    1  09:20:00   

     Franja Horaria                          Lugar Del Hecho Tipo De Calle  \
38               20               AUTOPISTA LUGONES PK 10000     AUTOPISTA   
106               0         AU BUENOS AIRES - LA PLATA KM. 4     AUTOPISTA   
138               0                     AUTOPISTA 25 DE MAYO     AUTOPISTA   
175               9            AV. LEOPOLDO LUGONES PKM 6900      GRAL PAZ   
179              11  AU PERITO MORENO Y RAMAL ENLACE AU1/AU6     AUTOPISTA   
180               3                       AU DELLEPIANE 2400     AUTOPISTA   
255              23              AU ARTURO FRONDIZI PKM 3100     AUTOPISTA   
312              22                 AUTOPISTA LUGONES KM 4.7     AUTOPISTA   
545               6             LUGONES, LEOPOLDO AV. KM 6,1     AUTOPISTA   
558              19                          MURGUIONDO 2700         CALLE   
620               9         AU BUENOS AIRES LA PLATA  KM 4,5     AUTOPISTA   

                                          Calle  Cruce  \
38                        LUGONES, LEOPOLDO AV.  False   
106           AUTOPISTA BUENOS AIRES - LA PLATA  False   
138                        AUTOPISTA 25 DE MAYO  False   
175                       LUGONES, LEOPOLDO AV.  False   
179                     AUTOPISTA PERITO MORENO  False   
180        AUTOPISTA DELLEPIANE LUIS TTE. GRAL.  False   
255  AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI  False   
312                       LUGONES, LEOPOLDO AV.  False   
545                       LUGONES, LEOPOLDO AV.  False   
558                                  MURGUIONDO  False   
620           AUTOPISTA BUENOS AIRES - LA PLATA  False   

                          Dirección Normalizada  Comuna    XY (CABA) Pos X  \
38                                           SD      13  Point (. .)     .   
106                                          SD       4  Point (. .)     .   
138                        AUTOPISTA 25 DE MAYO       0  Point (. .)     .   
175                       LUGONES, LEOPOLDO AV.      14  Point (. .)     .   
179                                          SD       9  Point (. .)     .   
180                                          SD       7  Point (. .)     .   
255  AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI       4  Point (. .)     .   
312                                          SD      14  Point (. .)     .   
545                                          SD      14  Point (. .)     .   
558                                  MURGUIONDO       9  Point (. .)     .   
620                                          SD       4  Point (. .)     .   

    Pos Y     Participantes Víctima      Acusado  
38      .           MOTO-SD    MOTO           SD  
106     .       MOTO-CARGAS    MOTO       CARGAS  
138     .             SD-SD      SD           SD  
175     .       MOTO-CARGAS    MOTO       CARGAS  
179     .       MOTO-CARGAS    MOTO       CARGAS  
180     .         AUTO-AUTO    AUTO         AUTO  
255     .    MOTO-PASAJEROS    MOTO    PASAJEROS  
312     .       PEATON-AUTO  PEATON         AUTO  
545     .  MOTO-OBJETO FIJO    MOTO  OBJETO FIJO  
558     .     PEATON-CA

In [34]:
hechos[hechos['XY (CABA)'] == 'Point (. .)']

Hechos_Id  Número De Victimas      Fecha   Año  Mes  Dia      Hora  \
35   2016-0049                   1 2016-04-17  2016    4   17  00:00:00   
38   2016-0052                   1 2016-04-20  2016    4   20  20:00:00   
71   2016-0096                   1 2016-07-25  2016    7   25  07:00:00   
106  2016-0136                   1 2016-10-25  2016   10   25  00:00:00   
138  2016-0174                   1 2016-12-27  2016   12   27  00:00:00   
175  2017-0042                   1 2017-04-10  2017    4   10  09:00:00   
179  2017-0050                   2 2017-04-28  2017    4   28  11:08:08   
180  2017-0051                   1 2017-05-01  2017    5    1  03:47:47   
255  2017-0140                   1 2017-11-19  2017   11   19  23:22:17   
312  2018-0039                   1 2018-04-21  2018    4   21  22:15:00   
545  2020-0026                   1 2020-05-17  2020    5   17  06:40:00   
558  2020-0039                   1 2020-09-01  2020    9    1  19:17:42   
620  2021-0023                   1 2021-03-01  2021    3    1  09:20:00   

     Franja Horaria                                    Lugar Del Hecho  \
35                0  AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI KM....   
38               20                         AUTOPISTA LUGONES PK 10000   
71                7       AUTOPISTA DELLEPIANE LUIS TTE. GRAL. KM. 2,3   
106               0                   AU BUENOS AIRES - LA PLATA KM. 4   
138               0                               AUTOPISTA 25 DE MAYO   
175               9                      AV. LEOPOLDO LUGONES PKM 6900   
179              11            AU PERITO MORENO Y RAMAL ENLACE AU1/AU6   
180               3                                 AU DELLEPIANE 2400   
255              23                        AU ARTURO FRONDIZI PKM 3100   
312              22                           AUTOPISTA LUGONES KM 4.7   
545               6                       LUGONES, LEOPOLDO AV. KM 6,1   
558              19                                    MURGUIONDO 2700   
620               9                   AU BUENOS AIRES LA PLATA  KM 4,5   

    Tipo De Calle                                       Calle  Cruce  \
35      AUTOPISTA  AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI  False   
38      AUTOPISTA                       LUGONES, LEOPOLDO AV.  False   
71      AUTOPISTA        AUTOPISTA DELLEPIANE LUIS TTE. GRAL.  False   
106     AUTOPISTA           AUTOPISTA BUENOS AIRES - LA PLATA  False   
138     AUTOPISTA                        AUTOPISTA 25 DE MAYO  False   
175      GRAL PAZ                       LUGONES, LEOPOLDO AV.  False   
179     AUTOPISTA                     AUTOPISTA PERITO MORENO  False   
180     AUTOPISTA        AUTOPISTA DELLEPIANE LUIS TTE. GRAL.  False   
255     AUTOPISTA  AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI  False   
312     AUTOPISTA                       LUGONES, LEOPOLDO AV.  False   
545     AUTOPISTA                       LUGONES, LEOPOLDO AV.  False   
558         CALLE                                  MURGUIONDO  False   
620     AUTOPISTA           AUTOPISTA BUENOS AIRES - LA PLATA  False   

                          Dirección Normalizada  Comuna    XY (CABA)  \
35   AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI       4  Point (. .)   
38                                           SD      13  Point (. .)   
71         AUTOPISTA DELLEPIANE LUIS TTE. GRAL.       8  Point (. .)   
106                                          SD       4  Point (. .)   
138                        AUTOPISTA 25 DE MAYO       0  Point (. .)   
175                       LUGONES, LEOPOLDO AV.      14  Point (. .)   
179                                          SD       9  Point (. .)   
180                                          SD       7  Point (. .)   
255  AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI       4  Point (. .)   
312                                          SD      14  Point (. .)   
545                                          SD      14  Point (. .)   
558                                  MURGUIO

Encontramos 13 filas que no presentan las coordenadas y también vemos 11 registros que tampoco presentan los datos de 'Pos X' y 'Pos Y'. Utilizando las herramientas de mapas de la web y la app intentaremos encontrar el mayor número de coordenadas posibles en función de la información proporcionada sobre la calle, dirección y ubicación de los eventos. Para las filas que no podamos completar se imputará 0 de tal forma que sirva para excluir el valor del tablero a realizar.

In [35]:
condition = (hechos['Lugar Del Hecho'] == 'AU BUENOS AIRES - LA PLATA KM. 4')
condition1 = (hechos['Lugar Del Hecho'] == 'AV. LEOPOLDO LUGONES PKM 6900')
condition2 = (hechos['Lugar Del Hecho'] == 'AU PERITO MORENO Y RAMAL ENLACE AU1/AU6')
condition3 = (hechos['Lugar Del Hecho'] == 'AU DELLEPIANE 2400')
condition4 = (hechos['Lugar Del Hecho'] == 'AU ARTURO FRONDIZI PKM 3100')
condition5 = (hechos['Lugar Del Hecho'] == 'AUTOPISTA LUGONES KM 4.7')
condition6 = (hechos['Lugar Del Hecho'] == 'LUGONES, LEOPOLDO AV. KM 6,1')
condition7 = (hechos['Lugar Del Hecho'] == 'MURGUIONDO 2700')
condition8 = (hechos['Lugar Del Hecho'] == 'AU BUENOS AIRES LA PLATA KM 4,5')


# Asignar valores a las columnas 'Pos Y' y 'Pos X' basándote en la condición
hechos.loc[condition, 'Pos Y'] = -34.635364
hechos.loc[condition, 'Pos X'] = -58.353722

hechos.loc[condition1, 'Pos Y'] = -34.54824546
hechos.loc[condition1, 'Pos X'] = -58.43854389

hechos.loc[condition2, 'Pos Y'] = -34.64061318
hechos.loc[condition2, 'Pos X'] = -58.49253465

hechos.loc[condition3, 'Pos Y'] = -34.66931534
hechos.loc[condition3, 'Pos X'] = -58.47557813

hechos.loc[condition4, 'Pos Y'] = -34.62832265
hechos.loc[condition4, 'Pos X'] = -58.37893323

hechos.loc[condition5, 'Pos Y'] = -34.54846070
hechos.loc[condition5, 'Pos X'] = -58.43846707

hechos.loc[condition6, 'Pos Y'] = -34.54116624
hechos.loc[condition6, 'Pos X'] = -58.45198104

hechos.loc[condition7, 'Pos Y'] = -34.66337786
hechos.loc[condition7, 'Pos X'] = -58.49320476

hechos.loc[condition8, 'Pos Y'] = -34.76823354
hechos.loc[condition8, 'Pos X'] = -58.17140557

In [36]:
hechos[['Pos X', 'Pos Y']] = hechos[['Pos X', 'Pos Y']].replace('.', 0)
hechos['XY (CABA)'] = hechos['XY (CABA)'].replace('Point (. .)', 0)
hechos[hechos['Pos X'] == 0]

Hechos_Id  Número De Victimas      Fecha   Año  Mes  Dia      Hora  \
38   2016-0052                   1 2016-04-20  2016    4   20  20:00:00   
138  2016-0174                   1 2016-12-27  2016   12   27  00:00:00   
620  2021-0023                   1 2021-03-01  2021    3    1  09:20:00   

     Franja Horaria                   Lugar Del Hecho Tipo De Calle  \
38               20        AUTOPISTA LUGONES PK 10000     AUTOPISTA   
138               0              AUTOPISTA 25 DE MAYO     AUTOPISTA   
620               9  AU BUENOS AIRES LA PLATA  KM 4,5     AUTOPISTA   

                                 Calle  Cruce Dirección Normalizada  Comuna  \
38               LUGONES, LEOPOLDO AV.  False                    SD      13   
138               AUTOPISTA 25 DE MAYO  False  AUTOPISTA 25 DE MAYO       0   
620  AUTOPISTA BUENOS AIRES - LA PLATA  False                    SD       4   

    XY (CABA) Pos X Pos Y Participantes Víctima Acusado  
38          0     0     0       MOTO-SD    MOTO      SD  
138         0     0     0         SD-SD      SD      SD  
620         0     0     0   MOTO-CARGAS    MOTO  CARGAS

Involucrado en los hechos

Ahora, por último analizaremos las columnas 'Participantes', 'Víctima' y 'Acusado'

In [37]:
hechos['Participantes'].unique()

array(['MOTO-AUTO', 'AUTO-PASAJEROS', 'MOTO-SD', 'MOTO-PASAJEROS',
       'MOTO-OBJETO FIJO', 'MOTO-CARGAS', 'PEATON-AUTO', 'PEATON-MOTO',
       'PEATON-PASAJEROS', 'AUTO-AUTO', 'AUTO-OBJETO FIJO',
       'PEATON-CARGAS', 'MOTO-MOTO', 'MULTIPLE', 'SD-SD', 'AUTO-CARGAS',
       'AUTO-SD', 'MOTO-MOVIL', 'CARGAS-CARGAS', 'PEATON-SD',
       'BICICLETA-CARGAS', 'SD-CARGAS', 'BICICLETA-PASAJEROS',
       'PASAJEROS-PASAJEROS', 'MOVIL-PASAJEROS', 'CARGAS-OBJETO FIJO',
       'PEATON-BICICLETA', 'SD-AUTO', 'CARGAS-AUTO', 'BICICLETA-AUTO',
       'MOTO-BICICLETA', 'SD-MOTO', 'MOVIL-CARGAS', 'PASAJEROS-AUTO',
       'AUTO-MOVIL', 'CARGAS-PASAJEROS', 'BICICLETA-TREN',
       'BICICLETA-OTRO', 'MOTO-OTRO', 'PEATON_MOTO-MOTO', 'PASAJEROS-SD'],
      dtype=object)

In [38]:
hechos['Víctima'].unique()

array(['MOTO', 'AUTO', 'PEATON', 'SD', 'CARGAS', 'BICICLETA', 'PASAJEROS',
       'MOVIL', 'OBJETO FIJO', 'PEATON_MOTO'], dtype=object)

In [39]:
hechos['Acusado'].unique()

array(['AUTO', 'PASAJEROS', 'SD', 'OBJETO FIJO', 'CARGAS', 'MOTO',
       'MULTIPLE', 'OTRO', 'BICICLETA', 'TREN'], dtype=object)

La columna 'Víctima' contiene los valores 'OBJETO FIJO' y 'PEATON_MOTO' que no están descritos en el diccionario de datos. Procedemos a observar estos valores.

In [40]:
hechos[hechos['Víctima'].isin(['OBJETO FIJO', 'PEATON_MOTO'])]

Hechos_Id  Número De Victimas      Fecha   Año  Mes  Dia      Hora  \
229  2017-0108                   2 2017-09-02  2017    9    2  04:53:08   
582  2020-0063                   2 2020-12-05  2020   12    5  07:10:00   

     Franja Horaria           Lugar Del Hecho Tipo De Calle           Calle  \
229               4   AV. GRAL. PAZ Y MACHAIN      GRAL PAZ  PAZ, GRAL. AV.   
582               7  NUEVA YORK Y ALTA GRACIA         CALLE      NUEVA YORK   

     Cruce     Dirección Normalizada  Comuna  \
229   True  PAZ, GRAL. AV. y MACHAIN      12   
582   True  NUEVA YORK y ALTA GRACIA      11   

                                  XY (CABA)         Pos X         Pos Y  \
229  Point (97098.48468623 109019.96106626)  -58.49491054  -34.54795581   
582  Point (94080.62190808 102083.62453795)  -58.52783814  -34.61047001   

        Participantes      Víctima Acusado  
229  AUTO-OBJETO FIJO  OBJETO FIJO    AUTO  
582  PEATON_MOTO-MOTO  PEATON_MOTO    MOTO

Vemos que en el primer caso los valores de las columnas 'Víctima' y 'Acusado' están invertidos. Por otro lado, es posible un pequeño error en la columna 'Participantes' con el valor 'PEATON_MOTO'. Procedemos a corregirlos:

In [41]:
# valor correcto en columna 'VICTIMA' y 'ACUSADO' en el indice 229
hechos.loc[[229], ['Víctima', 'Acusado']] = hechos.loc[[229], ['Acusado', 'Víctima']].values

# valor correcto en columna 'PARTICIPANTES' y 'VICTIMA' en el indice 582
hechos.loc[[582], ['Participantes', 'Víctima']] = hechos.loc[[582], ['Participantes', 'Víctima']].replace({'PEATON_MOTO-MOTO': 'PEATON-MOTO', 'PEATON_MOTO': 'PEATON'})
hechos[hechos['Hechos_Id'].isin(['2017-0108', '2020-0063'])]

Hechos_Id  Número De Victimas      Fecha   Año  Mes  Dia      Hora  \
229  2017-0108                   2 2017-09-02  2017    9    2  04:53:08   
582  2020-0063                   2 2020-12-05  2020   12    5  07:10:00   

     Franja Horaria           Lugar Del Hecho Tipo De Calle           Calle  \
229               4   AV. GRAL. PAZ Y MACHAIN      GRAL PAZ  PAZ, GRAL. AV.   
582               7  NUEVA YORK Y ALTA GRACIA         CALLE      NUEVA YORK   

     Cruce     Dirección Normalizada  Comuna  \
229   True  PAZ, GRAL. AV. y MACHAIN      12   
582   True  NUEVA YORK y ALTA GRACIA      11   

                                  XY (CABA)         Pos X         Pos Y  \
229  Point (97098.48468623 109019.96106626)  -58.49491054  -34.54795581   
582  Point (94080.62190808 102083.62453795)  -58.52783814  -34.61047001   

        Participantes Víctima      Acusado  
229  AUTO-OBJETO FIJO    AUTO  OBJETO FIJO  
582       PEATON-MOTO  PEATON         MOTO

Vamos a analizar columnas con valores 'SD'

In [42]:
ut.count_value_per_column(hechos, 'SD')

Cuenta de 'SD' en la columna Hechos_Id: 0
Cuenta de 'SD' en la columna Número De Victimas: 0
Cuenta de 'SD' en la columna Fecha: 0
Cuenta de 'SD' en la columna Año: 0
Cuenta de 'SD' en la columna Mes: 0
Cuenta de 'SD' en la columna Dia: 0
Cuenta de 'SD' en la columna Hora: 0
Cuenta de 'SD' en la columna Franja Horaria: 0
Cuenta de 'SD' en la columna Lugar Del Hecho: 0
Cuenta de 'SD' en la columna Tipo De Calle: 0
Cuenta de 'SD' en la columna Calle: 0
Cuenta de 'SD' en la columna Cruce: 0
Cuenta de 'SD' en la columna Dirección Normalizada: 7
Cuenta de 'SD' en la columna Comuna: 0
Cuenta de 'SD' en la columna XY (CABA): 0
Cuenta de 'SD' en la columna Pos X: 0
Cuenta de 'SD' en la columna Pos Y: 0
Cuenta de 'SD' en la columna Participantes: 0
Cuenta de 'SD' en la columna Víctima: 9
Cuenta de 'SD' en la columna Acusado: 22


In [43]:
hechos[hechos['Víctima'].isin(['SD'])]

Hechos_Id  Número De Victimas      Fecha   Año  Mes  Dia      Hora  \
35   2016-0049                   1 2016-04-17  2016    4   17  00:00:00   
76   2016-0101                   1 2016-08-07  2016    8    7  19:42:00   
88   2016-0115                   1 2016-09-02  2016    9    2  11:50:00   
92   2016-0119                   1 2016-09-04  2016    9    4  06:40:00   
138  2016-0174                   1 2016-12-27  2016   12   27  00:00:00   
163  2017-0029                   1 2017-03-07  2017    3    7  02:15:00   
200  2017-0074                   1 2017-06-04  2017    6    4  12:50:00   
213  2017-0089                   1 2017-07-13  2017    7   13  12:05:00   
268  2017-0155                   1 2017-12-12  2017   12   12  21:25:41   

     Franja Horaria                                    Lugar Del Hecho  \
35                0  AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI KM....   
76               19                         AV GRAL PAZ Y JORGE CHAVEZ   
88               11                                   ARIAS Y TRONADOR   
92                6              CRISOLOGO LARRALDE Y AVDA TRIUNVIRATO   
138               0                               AUTOPISTA 25 DE MAYO   
163               2                                    MOZART Y TANDIL   
200              12                                          CASCO 133   
213              12                          MARTINIANO LEGUIZAMON 901   
268              21                                 AV. RIVADAVIA 6350   

    Tipo De Calle                                       Calle  Cruce  \
35      AUTOPISTA  AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI  False   
76       GRAL PAZ                              PAZ, GRAL. AV.   True   
88          CALLE                                       ARIAS   True   
92        AVENIDA                     LARRALDE, CRISOLOGO AV.   True   
138     AUTOPISTA                        AUTOPISTA 25 DE MAYO  False   
163         CALLE                                      MOZART   True   
200         CALLE                                       CASCO  False   
213         CALLE                      LEGUIZAMON, MARTINIANO  False   
268       AVENIDA                               RIVADAVIA AV.  False   

                          Dirección Normalizada  Comuna  \
35   AUTOPISTA 1 SUR PRESIDENTE ARTURO FRONDIZI       4   
76               PAZ, GRAL. AV. y CHAVEZ, JORGE       9   
88                             ARIAS y TRONADOR      12   
92    LARRALDE, CRISOLOGO AV. y TRIUNVIRATO AV.      12   
138                        AUTOPISTA 25 DE MAYO       0   
163                             MOZART y TANDIL       9   
200                                   CASCO 133       9   
213                  LEGUIZAMON, MARTINIANO 901       9   
268                          RIVADAVIA AV. 6350       7   

                                   XY (CABA)               Pos X  \
35                                         0  -58.37714647568196   
76     Point (93946.81673714 97681.68907836)        -58.52932872   
88    Point (97608.87024259 108884.67029908)        -58.48935041   
92    Point (96944.69676081 107291.52292157)        -58.49659234   
138                                        0                   0   
163    Point (98009.35575364 97985.81952007)        -58.48501341   
200    Point (94281.31244858 99043.29027489)        -58.52567073   
213    Point (95246.78684010 97935.32361317)        -58.51514709   
268  Point (100674.98109200 100299.64949548)        -58.45593915   

                  Pos Y Participantes Víctima Acusado  
35   -34.63657525428238         SD-SD      SD      SD  
76         -34.65014993         SD-SD      SD      SD  
88         -34.54917673     SD-CARGAS      SD  CARGAS  
92         -34.56353635         SD-SD      SD      SD  
138                   0         SD-SD      SD      SD  
163        -34.64742455       SD-AUTO      SD    AUTO  
200        -34.63787791         SD-SD      SD      SD  
213        -34.64787051       SD-AUTO      SD    AUTO  
268        -34.6

No hay forma de transformar datos 'SD'

Ahora agregamos una nueva columna 'Día Semana'

In [44]:
locale.setlocale(locale.LC_TIME, 'es_ES')

# Añadimos la nueva columna 'Día semana'
hechos['Día Semana'] = hechos['Fecha'].dt.strftime('%A').str.capitalize()

hechos['Mes'] = hechos['Mes'].apply(lambda x: calendar.month_name[x]).str.capitalize()

# Restaurar a la configuración predeterminada
locale.setlocale(locale.LC_TIME, 'C')
# Visualizamos
hechos.head()

Hechos_Id  Número De Victimas      Fecha   Año    Mes  Dia      Hora  \
0  2016-0001                   1 2016-01-01  2016  Enero    1  04:00:00   
1  2016-0002                   1 2016-01-02  2016  Enero    2  01:15:00   
2  2016-0003                   1 2016-01-03  2016  Enero    3  07:00:00   
3  2016-0004                   1 2016-01-10  2016  Enero   10  00:00:00   
4  2016-0005                   1 2016-01-21  2016  Enero   21  05:20:00   

   Franja Horaria                            Lugar Del Hecho Tipo De Calle  \
0               4  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA   
1               1          AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ   
2               7                         AV ENTRE RIOS 2034       AVENIDA   
3               0      AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA   
4               5  AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA   

   ...  Cruce                              Dirección Normalizada Comuna  \
0  ...   True  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...      8   
1  ...   True               PAZ, GRAL. AV. y DE LOS CORRALES AV.      9   
2  ...  False                                ENTRE RIOS AV. 2034      1   
3  ...   True          LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.      8   
4  ...   True            SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.      1   

                                 XY (CABA)         Pos X         Pos Y  \
0    Point (98896.78238426 93532.43437792)  -58.47533969  -34.68757022   
1    Point (95832.05571093 95505.41641999)  -58.50877521  -34.66977709   
2   Point (106684.29090040 99706.57687843)  -58.39040293  -34.63189362   
3    Point (99840.65224780 94269.16534422)  -58.46503904  -34.68092974   
4  Point (106980.32827929 100752.16915795)  -58.38718297  -34.62246630   

    Participantes Víctima    Acusado Día Semana  
0       MOTO-AUTO    MOTO       AUTO    Viernes  
1  AUTO-PASAJEROS    AUTO  PASAJEROS     Sábado  
2       MOTO-AUTO    MOTO       AUTO    Domingo  
3         MOTO-SD    MOTO         SD    Domingo  
4  MOTO-PASAJEROS    MOTO  PASAJEROS     Jueves  

[5 rows x 21 columns]

In [45]:
def get_season(mes):
    if mes in [12, 1, 2]:
        return 'Verano'
    elif mes in [3, 4, 5]:
        return 'Otoño'
    elif mes in [6, 7, 8]:
        return 'Invierno'
    elif mes in [9, 10, 11]:
        return 'Primavera'
    else:
        return 'Desconocido'

# Agregamos nueva columna 'Estación'
hechos['Estación'] = hechos['Fecha'].dt.month.apply(get_season)
hechos

Hechos_Id  Número De Victimas      Fecha   Año        Mes  Dia      Hora  \
0    2016-0001                   1 2016-01-01  2016      Enero    1  04:00:00   
1    2016-0002                   1 2016-01-02  2016      Enero    2  01:15:00   
2    2016-0003                   1 2016-01-03  2016      Enero    3  07:00:00   
3    2016-0004                   1 2016-01-10  2016      Enero   10  00:00:00   
4    2016-0005                   1 2016-01-21  2016      Enero   21  05:20:00   
..         ...                 ...        ...   ...        ...  ...       ...   
690  2021-0093                   1 2021-12-13  2021  Diciembre   13  17:10:00   
691  2021-0094                   1 2021-12-20  2021  Diciembre   20  01:10:00   
692  2021-0095                   1 2021-12-30  2021  Diciembre   30  00:43:00   
693  2021-0096                   1 2021-12-15  2021  Diciembre   15  10:30:00   
694  2021-0097                   1 2021-11-18  2021  Noviembre   18  06:10:00   

     Franja Horaria                            Lugar Del Hecho Tipo De Calle  \
0                 4  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA   
1                 1          AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ   
2                 7                         AV ENTRE RIOS 2034       AVENIDA   
3                 0      AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA   
4                 5  AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA   
..              ...                                        ...           ...   
690              17                          AV. RIESTRA Y MOM       AVENIDA   
691               1                    AU DELLEPIANE Y LACARRA     AUTOPISTA   
692               0                        AV. GAONA Y TERRADA       AVENIDA   
693              10                         AV. EVA PERON 4071       AVENIDA   
694               6                    PADRE CARLOS MUJICA 709         CALLE   

     ...                              Dirección Normalizada  Comuna  \
0    ...  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...       8   
1    ...               PAZ, GRAL. AV. y DE LOS CORRALES AV.       9   
2    ...                                ENTRE RIOS AV. 2034       1   
3    ...          LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.       8   
4    ...            SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.       1   
..   ...                                                ...     ...   
690  ...                                  RIESTRA AV. y MOM       7   
691  ...         DELLEPIANE, LUIS, TTE. GRAL. y LACARRA AV.       9   
692  ...                                GAONA AV. y TERRADA      11   
693  ...                                PERON, EVA AV. 4071       9   
694  ...                            PADRE CARLOS MUGICA 709       1   

                                   XY (CABA)         Pos X         Pos Y  \
0      Point (98896.78238426 93532.43437792)  -58.47533969  -34.68757022   
1      Point (95832.05571093 95505.41641999)  -58.50877521  -34.66977709   
2     Point (106684.29090040 99706.57687843)  -58.39040293  -34.63189362   
3      Point (99840.65224780 94269.16534422)  -58.46503904  -34.68092974   
4    Point (106980.32827929 100752.16915795)  -58.38718297  -34.62246630   
..                                       ...           ...           ...   
690   Point (102728.60090138 98186.24929177)  -58.43353773  -34.64561636   
691    Point (99624.29795829 97569.69801131)  -58.46739825  -34.65117757   
692   Point (99116.45492358 101045.23284826)  -58.47293407  -34.61984745   
693    Point (99324.54463985 97676.26932409)  -58.47066794  -34.65021673   
694  Point (107664.16647795 104708.63962087)  -58.37976155  -34.58679619   

      Participantes    Víctima    Acusado Día Semana   Estación  
0         MOTO-AUTO       MOTO       AUTO    Viernes     Verano  
1    AUTO-PASAJEROS       AUTO  PASAJEROS     Sábado     Verano  
2         MOTO-AUTO       MOTO       AUTO    Domingo     Verano  
3           MOTO-SD       MOTO         SD

Ahora vamos a agregar la palabra 'Comuna' a los valores en la columna 'Comuna'

In [46]:
hechos["Comuna"] = hechos.apply(lambda x: "Comuna "+str(x["Comuna"]), axis=1)


Finalmente, reordenamos las columnas.

In [47]:
hechos.columns

Index(['Hechos_Id', 'Número De Victimas', 'Fecha', 'Año', 'Mes', 'Dia', 'Hora',
       'Franja Horaria', 'Lugar Del Hecho', 'Tipo De Calle', 'Calle', 'Cruce',
       'Dirección Normalizada', 'Comuna', 'XY (CABA)', 'Pos X', 'Pos Y',
       'Participantes', 'Víctima', 'Acusado', 'Día Semana', 'Estación'],
      dtype='object')

In [48]:
new_order = [
    'Hechos_Id', 'Número De Victimas', 'Fecha', 'Año', 'Mes', 'Dia', 'Día Semana', 'Estación',
    'Hora', 'Franja Horaria', 'Lugar Del Hecho', 'Tipo De Calle', 'Calle', 'Cruce', 'Dirección Normalizada',
    'Comuna', 'XY (CABA)', 'Pos X', 'Pos Y', 'Participantes', 'Víctima', 'Acusado'
]

# Reorder dataframe columns
hechos = hechos[new_order]
hechos.head(3)

Hechos_Id  Número De Victimas      Fecha   Año    Mes  Dia Día Semana  \
0  2016-0001                   1 2016-01-01  2016  Enero    1    Viernes   
1  2016-0002                   1 2016-01-02  2016  Enero    2     Sábado   
2  2016-0003                   1 2016-01-03  2016  Enero    3    Domingo   

  Estación      Hora  Franja Horaria  ...             Calle  Cruce  \
0   Verano  04:00:00               4  ...  PIEDRA BUENA AV.   True   
1   Verano  01:15:00               1  ...    PAZ, GRAL. AV.   True   
2   Verano  07:00:00               7  ...    ENTRE RIOS AV.  False   

                               Dirección Normalizada    Comuna  \
0  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...  Comuna 8   
1               PAZ, GRAL. AV. y DE LOS CORRALES AV.  Comuna 9   
2                                ENTRE RIOS AV. 2034  Comuna 1   

                                XY (CABA)         Pos X         Pos Y  \
0   Point (98896.78238426 93532.43437792)  -58.47533969  -34.68757022   
1   Point (95832.05571093 95505.41641999)  -58.50877521  -34.66977709   
2  Point (106684.29090040 99706.57687843)  -58.39040293  -34.63189362   

    Participantes Víctima    Acusado  
0       MOTO-AUTO    MOTO       AUTO  
1  AUTO-PASAJEROS    AUTO  PASAJEROS  
2       MOTO-AUTO    MOTO       AUTO  

[3 rows x 22 columns]

Guardamos nuestro df limpio en un archivo CSV

In [49]:
dataframe = [hechos]
name = ['hechos_limpio']
ut.save_to_csv(dataframe, name)

DataFrame 'hechos_limpio' guardado como 'data/csv/hechos_limpio.csv'
